### PR TITLE
Unite Dockerfiles - add build/run/push scripts

### PR DESCRIPTION
**Pull request by opadron**
In preparation for pushing Spack docker images in support of the Release and PR testing milestone, I observed that there was still a lot of duplication among the Dockerfiles for the various flavors of Linux, and there were still a lot of area for improvement.  This PR tries to cut duplication down as much as possible, and in a way that tries to be as maintainable as possible.

I've consolidated all the Dockerfiles into a single one.  It works using a shell script that wraps the script provided to docker during builds.  The script masks out subsets of the provided commands based on embedded annotations and forwards the result to `/bin/bash` for execution.  Using this script as the Dockerfile's `SHELL`, we are able to fully specify which commands to run in which Linux variants, and minimize duplication by providing commands common to multiple variants only once.

The reduction in duplication was so considerable that there was very little reason to continue requiring the use of `docker-compose`.  So, it was removed and replaced with shell scripts that make building predefined configurations with just plain docker very straightforward.

For example, to use the predefined configuration for ubuntu (based on bionic).

```Bash
cd share/spack/docker
source config/ubuntu.bash
./build-image.sh
# ... build ...
./run-image.sh --version # run "spack --version" inside the container
./run-image.sh # no arguments -- run an interactive shell session (with local tree mounted)
./push-image.sh # push to dockerhub (need permission to push to spack org)
```
In addition to making our docker offerings much simpler and more maintainable, this PR also cuts the guess work for building and publishing new versions down practically to zero.  Just source a config (or make your own), build, and push.  To support other variants of linux, just update the Dockerfile with the commands for your variant.

### CODE REVIEW

**Comment by opadron** [comment]
fixes #9200

**Comment by opadron** [comment]
Note to others: I've added changes to the `.travis.yml` file to try and get the ball rolling on automating the upload to dockerhub, but I am almost certain that my changes are not correct, so please have a look and let me know where I might need to make corrections.

File: .travis.yml [hunk 1]
```diff
@@ -108,13 +108,19 @@ jobs:
...
+      language: generic
+      env: TEST_SUITE=docker
   allow_failures:
     - dist: xenial
 
 stages:
   - 'style checks'
   - 'unit tests + documentation'
   - 'build tests'
+  - 'docker build'
```
Comment (tgamblin) [inline .travis.yml]: Does this need to run every time, on every PR?  It seems like that's too frequent.  I think it should only run on branch builds for develop -- so when `TRAVIS_BRANCH` is set to `develop`.

Comment (ax3l) [reply .travis.yml]: There is also a new `if` condition: https://blog.travis-ci.com/2017-09-12-build-stages-order-and-conditions
```
  - name: 'docker build'
    if: branch = develop```

**Comment by opadron** [comment]
Thanks for the help, @tgamblin, @ax3l.

I made the suggested change to the `.travis.yml` file.  Hopefully the images are built and pushed without issue.

While I was at it, I also made changes to how the all-in-one Dockerfile works.  As delightfully clever as I had thought my original solution was (using a custom SHELL), I decided to switch to preprocessing the Dockerfile, instead.  This change has several benefits:

 1. Changes to parts of the Dockerfile that are ultimately masked out for a build don't cause unnecessary  cache invalidation.  In other words: changing a part of the Dockerfile that is only for centos builds, for example, won't cause cache invalidation on ubuntu builds.  This change should reduce the number of unnecessary rebuild steps. 

 1. The new syntax for masking out portions of the Dockerfile should be more intuitive, and allows for masking on arbitrary conditions, rather than just the value of `$DISTRO`.

 1. A nice side effect of how the preprocessor reads in the Dockerfile is that the elements of the commands are canonicalized to a standard formatting prior to being passed along to docker.  This further reduces the number of unnecessary rebuilds due to changes in how a command is formatted from one version of the Dockerfile to the next.  For example: two RUN commands that differ only by insignificant whitespace should be seen as identical by docker.

At this point, best practices for working with Spack in docker are becoming sufficiently refined (and automated) that I wonder about adding a third evolution of this work in the form of new Spack commands!  I'm imagining being able to run our official builds as well as being able to roll your own custom builds and customize things like what compilers you want and what spack packages to preinstall, all using a simple command-line interface on a version of spack smart enough to do all the heavy lifting for you.

Anyway, this might be something I'd like to try at some point and I wonder what you and other folks think.

**Comment by tgamblin** [comment]
> At this point, best practices for working with Spack in docker are becoming sufficiently refined (and automated) that I wonder about adding a third evolution of this work in the form of new Spack commands! I'm imagining being able to run our official builds as well as being able to roll your own custom builds and customize things like what compilers you want and what spack packages to preinstall, all using a simple command-line interface on a version of spack smart enough to do all the heavy lifting for you.

> Anyway, this might be something I'd like to try at some point and I wonder what you and other folks think.

I think this is a really good idea.  If Spack could spit out Docker (or Singularity) containers automatically, that would be very cool.  I would wait on this feature until the very very near future -- I think that the changes I am about to push for Spack environments will make it easy to make the Spack portion of these docker files more declarative.

File: .travis.yml [hunk 2]
```diff
@@ -108,13 +108,20 @@ jobs:
...
+      env: TEST_SUITE=docker
   allow_failures:
     - dist: xenial
 
 stages:
   - 'style checks'
   - 'unit tests + documentation'
   - 'build tests'
+  - name: 'docker build'
+    if: type IN (api, cron, push)
```
Comment (tgamblin) [inline .travis.yml]: Do we want this to happen on `api` or `cron` builds?  `push` should be enough to handle each new commit on `develop`, right?  I think doing this on `api` is going to be bad, as I think that will only happen when we're testing something.  I could be wrong...

Comment (opadron) [reply .travis.yml]: I'm not certain how `api` and `cron` builds work in this context.  The particular change was in response to [this](https://github.com/spack/spack/pull/9329#discussion_r224696472) suggestion from the earlier review.  I'll list some alternatives below and you can just thumbs-up the one you prefer.

Comment (opadron) [reply .travis.yml]: Predicate on the branch name?

```Yaml
- name: 'docker build'
  if: branch = develop
```

Comment (opadron) [reply .travis.yml]: Predicate on the type (push only)?

```Yaml
- name: 'docker build'
  if: type = push
```

Comment (opadron) [reply .travis.yml]: Or, leave it as-is; predicate on the type (api, cron, and push)?

```Yaml
- name: 'docker build'
  if: type IN (api, cron, push)
```

Comment (tgamblin) [reply .travis.yml]: I think we just want pushes to develop and master for now.  I'll update it.

**Pull request merged**